### PR TITLE
Synchronize write mode among writers

### PR DESCRIPTION
**Pull request by arina-ielchiieva**
Currently Iceberg had three writers but they have different write modes:

Writer|Mode
-|-
Parquet|Overwrite
Avro|Create   
Orc|Create   

This PR synchronizes write modes in writers and defaults them to `create` as discussed in PR #302.

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
@rdblue please review.

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java [hunk 1]
```diff
@@ -64,6 +65,12 @@ private WriteBuilder(OutputFile file) {
       }
     }
 
+    public WriteBuilder forTable(Table table) {
+      schema(table.schema());
+      setAll(table.properties());
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/ORC.java]: I think this change is unrelated and should go in a separate PR. It's a good idea to add this, just not here.

Also, I think this should do some translation from table properties to ORC properties, like taking `write.orc.compression-codec` and setting the correct property for ORC.

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java [hunk 2]
```diff
@@ -40,7 +40,7 @@
   AvroFileAppender(Schema schema, OutputFile file,
                    Function<Schema, DatumWriter<?>> createWriterFunc,
                    CodecFactory codec, Map<String, String> metadata) throws IOException {
-    this.stream = file.create();
+    this.stream = file.createOrOverwrite();
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java]: This will affect writes for all data and metadata files. Is it safe to do this everywhere?

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 3]
```diff
@@ -210,7 +210,7 @@ private CompressionCodecName codec() {
             .setKeyValueMetadata(metadata)
             .setWriteSupport(getWriteSupport(type))
             .withCompressionCodec(codec())
-            .withWriteMode(ParquetFileWriter.Mode.OVERWRITE) // TODO: support modes
+            .withWriteMode(ParquetFileWriter.Mode.OVERWRITE)
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: I think this TODO item is still valid.

This comment refers to supporting modes passed into the Parquet write builder. That's different from #302 because it isn't a table property.

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java (hunk 2 above)
Comment (arina-ielchiieva) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java]: Moved it to Avro class and made configurable.

File: orc/src/main/java/org/apache/iceberg/orc/ORC.java (hunk 1 above)
Comment (arina-ielchiieva) [reply orc/src/main/java/org/apache/iceberg/orc/ORC.java]: Removed.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 3 above)
Comment (arina-ielchiieva) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Made write mode configurable via write builder, I guess this was meant by todo.

**Comment by arina-ielchiieva** [comment]
@rdblue thanks for the code review, addressed code review comments.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 3 above)
Comment (rdblue) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Great! That's exactly what the TODO meant.

File: core/src/main/java/org/apache/iceberg/avro/Avro.java [hunk 4]
```diff
@@ -124,6 +125,11 @@ public WriteBuilder meta(Map<String, String> properties) {
       return this;
     }
 
+    public WriteBuilder overwrite(boolean newOverwrite) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/Avro.java]: I suggest also adding `overwrite()` that sets `this.overwrite = true`. That way users can avoid passing a boolean constant.

File: core/src/main/java/org/apache/iceberg/ManifestListWriter.java [hunk 5]
```diff
@@ -73,6 +73,7 @@ public long length() {
           .schema(ManifestFile.schema())
           .named("manifest_file")
           .meta(meta)
+          .overwrite(false)
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/ManifestListWriter.java]: Let's default manifest lists and manifests to overwrite. These use UUID-based file names and should never conflict.

File: core/src/main/java/org/apache/iceberg/avro/Avro.java [hunk 6]
```diff
@@ -84,6 +84,7 @@ public static WriteBuilder write(OutputFile file) {
     private Map<String, String> config = Maps.newHashMap();
     private Map<String, String> metadata = Maps.newLinkedHashMap();
     private Function<Schema, DatumWriter<?>> createWriterFunc = GenericAvroWriter::new;
+    private boolean overwrite = true;
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/avro/Avro.java]: I think this should default to `false`.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 7]
```diff
@@ -138,6 +139,11 @@ public WriteBuilder metricsConfig(MetricsConfig newMetricsConfig) {
       return this;
     }
 
+    public WriteBuilder writeMode(ParquetFileWriter.Mode newWriteMode) {
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Instead of passing in a write mode, I think Parquet should use the same methods as the other file format helpers, `overwrite()` and `overwrite(boolean enabled)`. That way, all of them have a similar API and we don't leak Parquet classes through the Iceberg API.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java [hunk 8]
```diff
@@ -85,6 +85,7 @@ public static WriteBuilder write(OutputFile file) {
     private Map<String, String> config = Maps.newLinkedHashMap();
     private Function<MessageType, ParquetValueWriter<?>> createWriterFunc = null;
     private MetricsConfig metricsConfig = MetricsConfig.getDefault();
+    private ParquetFileWriter.Mode writeMode = ParquetFileWriter.Mode.OVERWRITE;
```
Comment (rdblue) [inline parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Can this default to non-overwrite mode?

File: core/src/main/java/org/apache/iceberg/ManifestListWriter.java (hunk 5 above)
Comment (arina-ielchiieva) [reply core/src/main/java/org/apache/iceberg/ManifestListWriter.java]: Done.

File: core/src/main/java/org/apache/iceberg/avro/Avro.java (hunk 6 above)
Comment (arina-ielchiieva) [reply core/src/main/java/org/apache/iceberg/avro/Avro.java]: Now all readers default to overwrite false.

File: core/src/main/java/org/apache/iceberg/avro/Avro.java (hunk 4 above)
Comment (arina-ielchiieva) [reply core/src/main/java/org/apache/iceberg/avro/Avro.java]: Added.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 8 above)
Comment (arina-ielchiieva) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Now all readers default to overwrite false.

File: parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java (hunk 7 above)
Comment (arina-ielchiieva) [reply parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java]: Agree, made the changes.

**Comment by rdblue** [comment]
Looks great! Thanks for fixing this, @arina-ielchiieva!
